util/timeutil: speedup Since

andreimatei opened the pull request:
Our implementation of timeutil.Since(t) was doing Now().Sub(t). Now,
it's using stdlib time.Since(). This is faster because, in the happy
path, the stdlib implementation only needs one of two system clock
reading (monotonic, wall clock). On my Linux machine this halves its
cost.

Release note: None

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71355#-:-MlpyNoWAeIsc3JGd2a2:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
There are a lot of `Now().Sub()` instances in our codebase, can you file an issue? We should also add a linter for that (I think there is one for the standard library `time.Now()` but not for our `timeutil.Now()`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71355#-:-MlqHM7vA_7rsLaW5_9c:b-wu0h8n)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
I've cleaned up all the `Now.Sub()` in the codebase and added a linter. 
And I've also done something else - I've introduced some new functions in `timeutil` that let the general public get their hands on a timestamp for which the optimized `time.Since(t)` code path actually kicks in. It turns out that it does not kick in for timestamps coming from `timeutil.Now()` because that does `Now().UTC()` and `UTC()` clears the monotonic measurement. I was interested in getting the optimization for the tracing package, and I happened to be getting it because the tracing package was exempt from the linter that enforced use of `timeutil.Now()` through some accident of history. But I want everybody to have fast `time.Since()`.
PTAL at the latter commits.

@dt you've introduced the UTC thing around https://github.com/cockroachdb/cockroach/pull/18783. I wonder if years later you still like it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71355#-:-Mlvlo0F9mlFy1PO8zUe:ba0ru3n)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten and @tbg)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
There is also the TimeSource / DefaultTimeSource. These are used in the tenant rate limiters (tenantrate, tenantcostclient). When writing the code, I kind of assumed the monotonic stuff was in use (CC @ajwerner). I think we should definitely consider removing our wrapper that does `UTC()` instead of adding more complication.

andreimatei committed:
util/timeutil: make our timestamps UTC in a different way

Before this patch, timeutil.Now() was doing time.Now().UTC(). This is in
contrast to stdlib time.Now(), which returns timestamps with the local
time zone. We have a linter against time.Now(), forcing use of
timeutil.Now(). The UTC-ization was done in #18783, which fixed #18774.
That issue had something to do with a timestamp leaking into a Datum
somehow and causing problems because it wasn't UTC.

The problem with UTC is that changing a timestamp's location strips the
timestamp's monotonic clock reading part [1]. So the timestamps we use
don't benefit from the properties that the monotonic reading is supposed
to give you and, more acutely, the lack of a monotonic reading makes
utilities like time.Since() twice as expensive as they need to be
because optimizations are disabled (on Linux you end up making two
syscalls instead of one - for reading two time sources).

This patch makes changes the way in which UTC timestamps are generated.
Instead of doing time.Now().UTC() every time, this patch does

  func init() {
    time.Local = nil  // nil is like time.UTC, but better
  }

Thus, we change the timezone to UTC, making all time.Now() calls return
UTC.

Even though it becomes less important to do so, this patch keeps the
linter prohibiting direct use of time.Now(), on the argument that we may
want to mock time in tests in the future, and so mandating everybody go
through our library keeps the door open to that.

Also in this patch, the tracing code is switched to using timeutil since
there's now no longer a performance penalty for it (before this patch,
through some accident of history, the tracing code was exempt from the
linter prohibiting use of time.Now().

Release note: The CRDB code now sets the time.Local variable to UTC.
This makes CRDB code even less friendly than it was to being used as a
library in other software projects.

[1] https://github.com/golang/go/blob/011fd002457da0823da5f06b099fcf6e21444b00/src/time/time.go#L208

Release note: None

andreimatei reviewed (commented):
I did the `init()` trick to set the local to UTC instead of what I had before. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71355#-:-Mm-6PiN-K7rILeEzWNs:b-fs9w63)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt and @nvanbenschoten)

___
*[pkg/util/timeutil/time.go, line 37 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/71355#-MlWxQdSXGXyPghUhtBT-r15-37:-MmThp2m9zLMGQkF4uu_:b-xwwg79) ([raw file](https://github.com/cockroachdb/cockroach/blob/577b8642a433925e179f8bfe9ac5a90e0ad823f2/pkg/util/timeutil/time.go#L37)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

what would you think of wrapping the underlying `time.Time` in a non-exported struct member, instead of just aliasing it, to force callers to use ToTime if they want at it?
</blockquote></details>

got rid of this


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 4 files at r1, 10 of 11 files at r18, 24 of 24 files at r19, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71355#-:-MmseLAn2F7M8AE5vSTB:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei and @dt)

___
*[-- commits, line 6 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/71355#-Mmscp9h2Tnic0WgEW1x):*
[nit] your English is poor like child
___
*[.bazelrc, line 7 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/71355#-MmsdRndC7Lv03sF1FWg:-MmsdRndC7Lv03sF1FWh:b-kai3op) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ff82015dd53ae5380ef0a82816a81c39e3a2db5/.bazelrc#L7)):*
> ```
> build:with_ui --define cockroach_with_ui=y
> build:test --define gotags=bazel,crdb_test,gss
> # Certain time based fail if UTC isn't the default timezone (see #25064).
> ```

[nit] Should this be above the line with `--test_env=TZ=`? Also, we should rephrase "certain time based fail" (here and in the Makefile)


<!-- Sent from Reviewable.io -->

andreimatei commented:
nothing to see here for now. Want to try another idea.
